### PR TITLE
[JBPM-9631] Exception generated on subprocess is not correctly propagated to parent process

### PR DESCRIPTION
**Pull request by elguardian**
Jira : https://issues.redhat.com/browse/JBPM-9631

### CODE REVIEW

File: jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        params.put("launchOnParent", Boolean.FALSE);
+
+        MutableBoolean error = new MutableBoolean(false);
+        kieSession.addEventListener(new DefaultProcessEventListener() {
+            @Override
+            public void afterNodeTriggered(ProcessNodeTriggeredEvent event) {
+                if(!"StartProcess".equals(event.getNodeInstance().getNodeName())) {
+                    return;
+                }
+                if(!"com.DealWithExceptionSubProcess".equals(event.getProcessInstance().getProcessId())) {
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java:53]: I think the issue is on the parent process, so we should add another listener or another test for the parent process to check that the exception is not `null`.
```suggestion
                if(!"com.DealWithException".equals(event.getProcessInstance().getProcessId())) {
```

File: jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    private static final String MAIN_PROCESS_ID = "com.DealWithException";
+
+    @Test
+    public void testMainSubprocessExceptionHandler() {
+        KieSession kieSession = createKSession(MAIN_PROCESS, MAIN_SUBPROCESS);
+        Map<String, Object> params = new HashMap<>();
+        params.put("launchOnParent", Boolean.FALSE);
+
+        MutableBoolean error = new MutableBoolean(false);
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java:46]: Wouldn't it be `AtomicBoolean` also valid here?

**Review by afalhambra — CHANGES_REQUESTED** [review]
Some changes required

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventSubProcessNodeInstance.java [hunk 3]
```diff
@@ -92,7 +93,8 @@ public void nodeInstanceCompleted(org.jbpm.workflow.instance.NodeInstance nodeIn
                     }
                     getProcessInstance().getMetaData().remove("SUB_PROCESS_INTERRUPTION");
                     if (getNodeInstanceContainer() instanceof ProcessInstance) {
-                        ((ProcessInstance) getProcessInstance()).setState(ProcessInstance.STATE_ABORTED, faultName);
+                        Object faultData = getProcessInstance().getVariable("subexception");
```
Comment (albfan) [inline jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventSubProcessNodeInstance.java:96]: I replace it with:

    Object faultData = getProcessInstance().getVariable("event");
    
and seems to be enough. `event` is the general variable you get capturing an error

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventSubProcessNodeInstance.java [hunk 4]
```diff
@@ -92,7 +93,8 @@ public void nodeInstanceCompleted(org.jbpm.workflow.instance.NodeInstance nodeIn
...
+                        ((ProcessInstance) getProcessInstance()).setState(ProcessInstance.STATE_ABORTED, faultName, faultData);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventSubProcessNodeInstance.java:97]: done

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventSubProcessNodeInstance.java [hunk 5]
```diff
@@ -23,6 +23,7 @@
 import org.jbpm.workflow.core.node.EventSubProcessNode;
 import org.jbpm.workflow.core.node.StartNode;
 import org.jbpm.workflow.instance.NodeInstanceContainer;
+import org.jbpm.workflow.instance.impl.WorkflowProcessInstanceImpl;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventSubProcessNodeInstance.java:26]: done

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/SubProcessNodeInstance.java [hunk 6]
```diff
@@ -334,12 +334,12 @@ public void processInstanceCompleted(ProcessInstance processInstance) {
 
     }
 
-    private boolean handleError(String faultName, ProcessInstance processInstance) {
+    private boolean handleError(Object faultData, String faultName, ProcessInstance processInstance) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/SubProcessNodeInstance.java:337]: done

File: jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java:2]: done

File: jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java (hunk 2 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java:46]: not needed.

File: jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                    return;
+                }
+
+                Exception e = (Exception) ((WorkflowProcessInstanceImpl) event.getProcessInstance()).getVariable("event");
+                error.setValue(e != null);
+            }
+        });
+        ProcessInstance pi = kieSession.startProcess(MAIN_PROCESS_ID, params);
+        this.assertProcessInstanceAborted(pi.getId());
+        Assert.assertTrue(error.getValue());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java:63]: done

File: jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java (hunk 1 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/regression/subprocess/MainProcessExceptionHandlerTest.java:53]: correct. that is the reason we are looking into it.

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventSubProcessNodeInstance.java (hunk 3 above)
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventSubProcessNodeInstance.java:96]: right

[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/EventSubProcessNodeInstance.java:96]: it should be event

**Review by afalhambra — CHANGES_REQUESTED** [review]
Unit test should be changed IMO

**Comment by elguardian** [comment]
Jenkins retest this please

**Review by afalhambra — APPROVED** [review]
Awesome work @elguardian! Thanks!

**Comment by elguardian** [comment]
Jenkins retest this
